### ISSUE TITLE
Remote Firefox in xvfb hangs exactly on the second test

### ISSUE THREAD

**Issue by Nakilon** [jnicklas/capybara#1586]
I run tests with Capybara -> WebdriverRemote -> Firefox -> xvfb launching them as:
`java -jar selenium-2.47.1.jar -role hub`
`java -jar selenium-2.47.1.jar -role webdriver -hubHost 127.0.0.1`
They are hanging on `visit` starting exactly from **the second test** in 10% of test runs.

![screen shot 2015-10-08 at 13 47 01](https://cloud.githubusercontent.com/assets/2870363/10427816/5adf8426-70f5-11e5-91c3-40179b4cde72.png)

Here is node log around the moment of hanging:
```
...
12:57:02.307 INFO - Done: [get current url]
12:57:02.309 INFO - Executing: [get current url])
12:57:02.319 INFO - Done: [get current url]
12:57:02.321 INFO - Executing: [delete all cookies])
12:57:02.330 INFO - Done: [delete all cookies]
12:57:02.333 INFO - Executing: [get: about:blank])
12:57:02.544 INFO - Done: [get: about:blank]
12:57:02.547 INFO - Executing: [find elements: By.xpath: /html/body/*])
12:57:02.556 INFO - Done: [find elements: By.xpath: /html/body/*]
12:57:02.559 INFO - Executing: [get: https://my.example.com/])
12:58:02.620 INFO - Executing: [get current url])
12:59:02.680 INFO - Executing: [get current url])
13:00:02.741 INFO - Executing: [delete all cookies])
13:01:02.802 INFO - Executing: [delete all cookies])
13:02:02.847 INFO - Executing: [get: https://my.example.com/])
13:03:02.907 INFO - Executing: [get current url])
13:04:02.969 INFO - Executing: [get current url])
13:05:03.029 INFO - Executing: [delete all cookies])
13:06:03.069 INFO - Executing: [delete all cookies])
13:07:03.131 INFO - Executing: [get: https://my.example.com/])
13:08:03.147 INFO - Executing: [get current url])
13:09:03.194 INFO - Executing: [get current url])
13:10:03.200 INFO - Executing: [delete all cookies])
13:11:03.245 INFO - Executing: [delete all cookies])
...
```
... and on suite finish:
```
...
13:17:03.458 INFO - Executing: [delete session: f613a69a-b4fe-42b3-b75b-8824eca064d0])
13:18:03.518 INFO - Executing: [delete session: f613a69a-b4fe-42b3-b75b-8824eca064d0])
13:27:02.781 WARN - Exception thrown
org.openqa.selenium.TimeoutException: Timed out waiting for page load.
Command duration or timeout: 1800.01 seconds
Build info: version: '2.47.1', revision: '411b314', time: '2015-07-30 03:03:16'
System info: host: 'localhost', ip: '127.0.0.1', os.name: 'Linux', os.arch: 'amd64', os.version: '2.6.32-042stab108.1', java.version: '1.8.0_51'
Driver info: org.openqa.selenium.firefox.FirefoxDriver
Capabilities [{applicationCacheEnabled=true, rotatable=false, handlesAlerts=true, databaseEnabled=true, version=38.3.0, platform=LINUX, nativeEvents=false, acceptSslCerts=true, webStorageEnabled=true, locationContextEnabled=true, browserName=firefox, takesScreenshot=true, javascriptEnabled=true, cssSelectorsEnabled=true}]
Session ID: 906d4a85-36bf-4637-b3c3-cfec0cd2ad49
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
        at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
        at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:595)
        at org.openqa.selenium.remote.RemoteWebDriver.get(RemoteWebDriver.java:306)
        at sun.reflect.GeneratedMethodAccessor33.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
...
        at org.openqa.selenium.support.events.EventFiringWebDriver$2.invoke(EventFiringWebDriver.java:102)
        at com.sun.proxy.$Proxy1.get(Unknown Source)
        at org.openqa.selenium.support.events.EventFiringWebDriver.get(EventFiringWebDriver.java:160)
        at org.openqa.selenium.remote.server.handler.ChangeUrl.call(ChangeUrl.java:40)
        at org.openqa.selenium.remote.server.handler.ChangeUrl.call(ChangeUrl.java:1)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.openqa.selenium.remote.server.DefaultSession$1.run(DefaultSession.java:176)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.openqa.selenium.TimeoutException: Timed out waiting for page load.
Build info: version: '2.47.1', revision: '411b314', time: '2015-07-30 03:03:16'
System info: host: 'localhost', ip: '127.0.0.1', os.name: 'Linux', os.arch: 'amd64', os.version: '2.6.32-042stab108.1', java.version: '1.8.0_51'
Driver info: driver.version: unknown
        at <anonymous class>.Utils.initWebLoadingListener/<(file:///tmp/anonymous1095388925762721462webdriver-profile/extensions/fxdriver@googlecode.com/components/driver-component.js:8993)
        at <anonymous class>.WebLoadingListener/e(file:///tmp/anonymous1095388925762721462webdriver-profile/extensions/fxdriver@googlecode.com/components/driver-component.js:5076)
        at <anonymous class>.WebLoadingListener/<(file:///tmp/anonymous1095388925762721462webdriver-profile/extensions/fxdriver@googlecode.com/components/driver-component.js:5084)
        at <anonymous class>.fxdriver.Timer.prototype.setTimeout/<.notify(file:///tmp/anonymous1095388925762721462webdriver-profile/extensions/fxdriver@googlecode.com/components/driver-component.js:623)
13:27:02.782 WARN - Exception: Timed out waiting for page load.
...
```
Note 0: it doesn't produce screenshots.
Note 1: that the same `visit` works fine in the first test:
```
...
12:56:55.646 INFO - Executing: [get: https://my.example.com/])
12:56:58.281 INFO - Done: [get: https://my.example.com/]
...
```
Note 2: the only difference of the second test from the first one I know is that it already has Facebook cookies -- authorizing like this:
```
RSpec.configure do |c|
  c.before do |example|
    visit "/"
    find(:xpath, "//a[contains(., 'Log in with Facebook')]").click
    if current_host == "https://www.facebook.com"
      find("#email").set "..."
      find("#pass").set "...\n"
      Selenium::WebDriver::Wait.new().until do
        current_host == Capybara.app_host
      end
      expect(page).to have_no_selector(".page_preloader", wait: 15)
    end
  end
end
```
Here are parts of my Gemfile.lock:
```
    capybara (2.5.0)
    capybara-screenshot (1.0.11)
      capybara (>= 1.0, < 3)
    rspec (3.3.0)
      rspec-core (~> 3.3.0)
      rspec-expectations (~> 3.3.0)
      rspec-mocks (~> 3.3.0)
    rspec-core (3.3.2)
      rspec-support (~> 3.3.0)
    rspec-expectations (3.3.1)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.3.0)
    rspec-mocks (3.3.2)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.3.0)
    rspec-support (3.3.0)
    rspec_junit_formatter (0.2.3)
      builder (< 4)
      rspec-core (>= 2, < 4, != 2.12.0)
    selenium-webdriver (2.47.1)
      childprocess (~> 0.5)
      multi_json (~> 1.0)
      rubyzip (~> 1.0)
      websocket (~> 1.0)

PLATFORMS
  ruby

DEPENDENCIES
  capybara-screenshot
  rspec
  rspec_junit_formatter
  selenium-webdriver

BUNDLED WITH
   1.10.6
```
CentOS release 6.7 (Final), Mozilla Firefox 38.3.0 and openjdk version "1.8.0_51".

**Comment by twalpole** [comment]
Since all capybara is doing is requesting selenium-webdriver to visit a url I would guess you've run into a bug in selenium and should report it over on that project.

I do have a concern about your test code though - in your before block the 

    if current_host == "https://www.facebook.com"

there is no guarantee the previous click has actually changed the current_host immediately so it's possible that equality test is failing when you expected it to pass - you would probably be better off using something like

    if page.has_current_path?(/http:\/\/www.facebook.com/, url: true)

which will wait a bit to see if the url changes, you can also use similar rather than directly using Selenium::WebDriver::Wait which utilizes Time.now and can therefore hang if Time is frozen/stubbed/etc  - On platforms that support it capybara utilizes the monotonic system timer so doesn't have that issue.

**Comment by Nakilon** [comment]
Seems to be fixed by moving from `selenium-2.47.1.jar` to `selenium-2.48.2.jar` and updating the gem to `selenium-webdriver (2.48.0)`.
